storage/src/journal/contiguous/mod.rs

```diff
@@ -0,0 +1,143 @@
+//! A contiguous journal interface for position-based append-only logging.
+//!
+//! This module includes:
+//!
+//! - [Contiguous]: Trait for append-only log
+//! - [Variable]: Wrapper for [super::variable::Journal] that implements [Contiguous]
+
+use super::Error;
+use futures::Stream;
+use std::num::NonZeroUsize;
+
+mod fixed;
+mod variable;
+
+// Re-export public types
+pub use variable::{Config, Variable};
+
+#[cfg(test)]
+pub(super) mod tests;
+
+/// Core trait for contiguous journals supporting sequential append operations.
+///
+/// A contiguous journal maintains a monotonically increasing position counter where each
+/// appended item receives a unique position starting from 0.
+pub trait Contiguous {
+    /// The type of items stored in the journal.
+    type Item;
+
+    /// Append a new item to the journal, returning its position.
+    ///
+    /// Positions are monotonically increasing starting from 0. The position of each item
+    /// is stable across pruning (i.e., if item X has position 5, it will always have
+    /// position 5 even if earlier items are pruned).
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the underlying storage operation fails or if the item cannot
+    /// be encoded.
+    fn append(
+        &mut self,
+        item: Self::Item,
+    ) -> impl std::future::Future<Output = Result<u64, Error>> + Send;
+
+    /// Return the total number of items that have been appended to the journal.
+    ///
+    /// This count is NOT affected by pruning. The next appended item will receive this
+    /// position as its value.
+    fn size(&self) -> impl std::future::Future<Output = Result<u64, Error>> + Send;
+
+    /// Return the position of the oldest item still retained in the journal.
+    ///
+    /// Returns `None` if the journal is empty or if all items have been pruned.
+    ///
+    /// After pruning, this returns the position of the first item that remains.
+    /// Note that due to section/blob alignment, this may be less than the `min_position`
+    /// passed to `prune()`.
+    fn oldest_retained_pos(
+        &self,
+    ) -> impl std::future::Future<Output = Result<Option<u64>, Error>> + Send;
+
+    /// Prune items at positions strictly less than `min_position`.
+    ///
+    /// Returns `true` if any data was pruned, `false` otherwise.
+    ///
+    /// # Behavior
+    ///
+    /// - If `min_position > size()`, the prune is capped to `size()` (no error is returned)
```

Comment:
Should we document the atomicity property here? (e.g. that it's not atomic, but if it fails, there won't be gaps left)?